### PR TITLE
L2 Emergency Protocol (Arbitrum) update

### PR DESCRIPTION
**Pull request by krebernisak**
- critical Arbitrum network update, xDomain `msg.sender` backwards incompatible change

### CODE REVIEW

File: contracts/src/v0.8/dev/ArbitrumValidator.sol [hunk 1]
```diff
@@ -240,41 +286,48 @@ contract ArbitrumValidator is TypeAndVersionInterface, AggregatorValidatorInterf
...
+    address target = L2_FLAGS;
+    // Choose and encode the underlying Flags call
+    bytes memory data = currentAnswer == ANSWER_SEQ_OFFLINE ? CALL_RAISE_FLAG : CALL_LOWER_FLAG;
+    bytes memory message = abi.encodeWithSelector(selector, target, data);
+    // Make the xDomain call
+    // NOTICE: if gasCostL2Value is zero the payment is processed on L2 so the L2 address needs to be funded, as it will
     // paying the fee. We also ignore the returned msg number, that can be queried via the InboxMessageDelivered event.
-    s_inbox.createRetryableTicket{value: s_gasConfig.gasCostL2}(
-      s_l2FlagsAddress,
+    IInbox(CROSS_DOMAIN_MESSENGER).createRetryableTicketNoRefundAliasRewrite{value: s_gasConfig.gasCostL2Value}(
```
Comment (RodrigoAD) [inline contracts/src/v0.8/dev/ArbitrumValidator.sol:298]: `createRetryableTicketNoRefundAliasRewrite` I imagine doesn't change `refundAddr` address to the new L2 alias right?

Comment (krebernisak) [reply contracts/src/v0.8/dev/ArbitrumValidator.sol:298]: Yes, exactly! The old `createRetryableTicket` also does alias the `refundAddr` addr, so we use this one instead to keep it straightforward.

**Comment by Mint-6263** [comment]
Ok

File: contracts/src/v0.8/dev/ArbitrumValidator.sol [hunk 2]
```diff
@@ -302,6 +295,8 @@ contract ArbitrumValidator is TypeAndVersionInterface, AggregatorValidatorInterf
       return true;
     }
 
+    // Excess gas on L2 will be sent to the L2 xDomain alias address of this contract
+    address refundAddr = AddressAliasHelper.applyL1ToL2Alias(address(this));
```
Comment (RodrigoAD) [inline contracts/src/v0.8/dev/ArbitrumValidator.sol:299]: At the end we are doing what `createRetryableTicket` does for us. If refundAddr will always be this contract aliased, we can let Arbitrum do the aliasing for us

Comment (krebernisak) [reply contracts/src/v0.8/dev/ArbitrumValidator.sol:299]: That's true, but we'd need to introduce some trust as they have a `shouldRewriteSender` toggle available: https://github.com/OffchainLabs/arbitrum/blob/38e1600c30ef64cba8e52dbafe25af80f2591d38/packages/arb-bridge-eth/contracts/bridge/Inbox.sol#L337

That's the reason I've decided to take control of the aliasing.

Comment (RodrigoAD) [reply contracts/src/v0.8/dev/ArbitrumValidator.sol:323]: Would be nice to have some way to inspect this address now that is not on GasConfig. As it's a constant now, could we do the same as other constants and make it `address immutable public`?

File: contracts/src/v0.8/dev/ArbitrumValidator.sol [hunk 3]
```diff
@@ -240,52 +324,111 @@ contract ArbitrumValidator is TypeAndVersionInterface, AggregatorValidatorInterf
...
-      s_gasConfig.maxGasPrice,
-      currentAnswer == isServiceOffline ? CALL_RAISE_FLAG : CALL_LOWER_FLAG
+      maxSubmissionCost,
+      refundAddr, // excessFeeRefundAddress
+      refundAddr, // callValueRefundAddress
+      maxGas,
+      gasPriceBid,
+      message
     );
+    // return success
```
Comment (RodrigoAD) [inline contracts/src/v0.8/dev/ArbitrumValidator.sol:353]: Would an event be useful here? As on `withdrawFundsFromL2` with `emit L2WithdrawalRequested(id, amount, refundAddr);`

Comment (krebernisak) [reply contracts/src/v0.8/dev/ArbitrumValidator.sol:353]: Crossed my mind, but potentially an overkill to add another event between the aggregator's `Transmit` one and the bridge `InboxMessageDelivered` (`SentMessage` on Optimism). Tracking `Transmit` events should give us all the required info.

Do you have a use case in mind?

Comment (RodrigoAD) [reply contracts/src/v0.8/dev/ArbitrumValidator.sol:353]: The only case I was thinking it was tracking the message, but `InboxMessageDelivered` does it already then

**Review by RodrigoAD — APPROVED** [review]

**Pull request merged**